src/Http/Http/src/Internal/ItemsDictionary.cs

```diff
@@ -8,111 +8,147 @@ namespace Microsoft.AspNetCore.Http.Internal
 {
     public class ItemsDictionary : IDictionary<object, object>
     {
+        private IDictionary<object, object> _items;
+
         public ItemsDictionary()
-            : this(new Dictionary<object, object>())
-        {
-        }
+        {}
 
         public ItemsDictionary(IDictionary<object, object> items)
         {
-            Items = items;
+            _items = items;
         }
 
-        public IDictionary<object, object> Items { get; }
+        public IDictionary<object, object> Items => this;
 
         // Replace the indexer with one that returns null for missing values
         object IDictionary<object, object>.this[object key]
         {
             get
             {
-                object value;
-                if (Items.TryGetValue(key, out value))
+                if (_items != null && _items.TryGetValue(key, out var value))
                 {
                     return value;
                 }
                 return null;
             }
-            set { Items[key] = value; }
+            set
+            {
+                EnsureDictionary();
+                _items[key] = value;
+            }
         }
 
         void IDictionary<object, object>.Add(object key, object value)
         {
-            Items.Add(key, value);
+            EnsureDictionary();
+            _items.Add(key, value);
         }
 
         bool IDictionary<object, object>.ContainsKey(object key)
-        {
-            return Items.ContainsKey(key);
-        }
+            => _items != null && _items.ContainsKey(key);
 
         ICollection<object> IDictionary<object, object>.Keys
         {
-            get { return Items.Keys; }
+            get
+            {
+                EnsureDictionary();
+                return _items.Keys;
+            }
         }
 
         bool IDictionary<object, object>.Remove(object key)
-        {
-            return Items.Remove(key);
-        }
+            => _items != null && _items.Remove(key);
 
         bool IDictionary<object, object>.TryGetValue(object key, out object value)
         {
-            return Items.TryGetValue(key, out value);
+            value = null;
+            return _items != null && _items.TryGetValue(key, out value);
         }
 
         ICollection<object> IDictionary<object, object>.Values
         {
-            get { return Items.Values; }
+            get
+            {
+                EnsureDictionary();
```

Comment:
Why allocate here if the inner dictionary is null?

Reply:
Mistakenly thought the collections were mutable

Reply:
Changed